fluent-dom/src/document_localization.js

```diff
@@ -0,0 +1,179 @@
+import Localization from './localization';
+
+export default class DocumentLocalization extends Localization {
+  constructor(doc, resIds, generateMessage) {
+    super(doc.location.href, resIds, generateMessage);
+    this.document = doc;
+    this.query = '[data-l10n-id]';
+    this.roots = new Set();
+    this.mutationObserver = new doc.defaultView.MutationObserver(
+      mutations => this.translateMutations(mutations)
+    );
+
+    this.observerConfig = {
+      attribute: true,
+      characterData: false,
+      childList: true,
+      subtree: true,
+      attributeFilter: ['data-l10n-id', 'data-l10n-args']
+    };
+  }
+
+  onLanguageChange() {
+    super.onLanguageChange();
+    this.translateDocument();
+  }
+
+  setAttributes(element, id, args) {
+    element.setAttribute('data-l10n-id', id);
+    if (args) {
+      element.setAttribute('data-l10n-args', JSON.stringify(args));
+    }
+    return element;
+  }
+
+  getAttributes(element) {
+    return {
+      id: element.getAttribute('data-l10n-id'),
+      args: JSON.parse(element.getAttribute('data-l10n-args'))
+    };
+  }
+
+  connectRoot(root) {
+    this.roots.add(root);
+    this.observeRoot(root);
+  }
+
+  disconnectRoot(root) {
+    this.roots.delete(root);
+    this.unobserveRoot(root);
+
+    return this.roots.size === 0;
+  }
+
+  translateRoots() {
+    const roots = Array.from(this.roots);
+    return Promise.all(
+      roots.map(root => this.translateRoot(root))
+    );
+  }
+
+  translateRoot(root) {
+    return this.translateFragment(root);
```

Comment:
Can you elaborate why? Setting the attributes is essential for the completeness of the localized DOM. Dispatching the event is harmless in the inert scenario (build-time or server-side rendering); moving it to `fluent-web` will mix concerns.